Source/WebKit/ModelProcess/ModelProcess.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ModelProcess.h"
+
+#if ENABLE(MODEL_PROCESS)
+
+#include "ArgumentCoders.h"
+#include "Attachment.h"
+#include "AuxiliaryProcessMessages.h"
+#include "DataReference.h"
+#include "LogInitialization.h"
+#include "ModelConnectionToWebProcess.h"
+#include "ModelProcessConnectionParameters.h"
+#include "ModelProcessCreationParameters.h"
+#include "ModelProcessProxyMessages.h"
+#include "WebPageProxyMessages.h"
+#include "WebProcessPoolMessages.h"
+#include <WebCore/CommonAtomStrings.h>
+#include <WebCore/LogInitialization.h>
+#include <WebCore/MemoryRelease.h>
+#include <wtf/Algorithms.h>
+#include <wtf/CallbackAggregator.h>
+#include <wtf/LogInitialization.h>
+#include <wtf/MemoryPressureHandler.h>
+#include <wtf/OptionSet.h>
+#include <wtf/RunLoop.h>
+#include <wtf/Scope.h>
+#include <wtf/UniqueRef.h>
+#include <wtf/text/AtomString.h>
+
+namespace WebKit {
+
+// We wouldn't want the ModelProcess to repeatedly exit then relaunch when under memory pressure. In particular, we need to make sure the
+// WebProcess has a chance to schedule work after the ModelProcess gets launched. For this reason, we make sure that the ModelProcess never
+// idle-exits less than 5 seconds after getting launched. This amount of time should be sufficient for the WebProcess to schedule
+// work in the ModelProcess.
+constexpr Seconds minimumLifetimeBeforeIdleExit { 5_s };
```

Comment:
This has a lot of duplicate code with the GPU process.  It would be much better to share code by maybe inheriting from a shared superclass rather than duplicate code.

Reply:
This is indeed a feedback from a number of engineers too. We are going to explore that in the near future.